src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1419,6 +1662,271 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
+                {
+                    moveShard(*prev);
+                }
+
+                prev = boost::none;
+            }
+
+            if (cur)
+            {
+                // The formerly latest shard is now
+                // the second latest
+                if (cur == latestShardIndex - 1)
+                {
+                    prev = cur;
+                }
+
+                // The formerly latest shard is no
+                // longer a 'recent' shard
+                else
+                {
+                    // Move the formerly latest shard to
+                    // historical storage
+                    if (keepShard(*cur) && separateHistoricalPath)
+                    {
+                        moveShard(*cur);
+                    }
+                }
+
+                cur = boost::none;
+            }
+        }
+    }
+}
+
+auto
+DatabaseShardImp::expectNewShard(
+    std::uint32_t shardIndex,
+    std::uint32_t numHistoricalShards,
+    std::lock_guard<std::mutex> const& lock) -> boost::optional<PathDesignation>
+{
+    // Any shard earlier than the two most recent shards
+    // is a historical shard
+    auto const boundaryIndex = getShardBoundaryIndex(lock);
+    auto const isHistoricalShard = shardIndex < boundaryIndex;
+
+    auto const designation = isHistoricalShard && !historicalPaths_.empty()
+        ? PathDesignation::historical
+        : PathDesignation::none;
+
+    // Check shard count and available storage space
+    if (isHistoricalShard && numHistoricalShards >= maxHistoricalShards_)
+    {
+        JLOG(j_.error()) << "maximum number of historical shards reached";
+        canAdd_ = false;
+        return boost::none;
+    }
+    if (!sufficientStorage(1, designation, lock))
+    {
+        JLOG(j_.error()) << "insufficient storage space available";
+        canAdd_ = false;
+        return boost::none;
+    }
+
+    return designation;
+}
+
+boost::filesystem::path
+DatabaseShardImp::chooseHistoricalPath()
+{
+    std::lock_guard lock(mutex_);
+    return chooseHistoricalPath(lock);
+}
+
+boost::filesystem::path
+DatabaseShardImp::chooseHistoricalPath(std::lock_guard<std::mutex> const&)
+{
+    // If not configured with separate historical paths,
+    // use the main path (dir_) by default.
+    if (historicalPaths_.empty())
+        return dir_;
+
+    boost::filesystem::path historicalShardPath;
+    std::vector<boost::filesystem::path> potentialPaths;
+
+    for (boost::filesystem::path const& path : historicalPaths_)
+    {
+        if (boost::filesystem::space(path).available >= avgShardFileSz_)
+            potentialPaths.push_back(path);
+    }
+
+    if (potentialPaths.empty())
+    {
+        JLOG(j_.error()) << "failed to select a historical shard path";
+        return "";
+    }
+
+    std::sample(
+        potentialPaths.begin(),
+        potentialPaths.end(),
+        &historicalShardPath,
+        1,
+        default_prng());
```

Comment:
Instead of a random sample, why not choose the one with the most space available?

Reply:
It's not necessary - see [this comment thread](https://github.com/ripple/rippled/pull/3463#discussion_r447239944).